### ISSUE TITLE
Alignment and infra geom authoring experience in BlenderBIM

### ISSUE THREAD

**Issue by aothms** [IfcOpenShell/IfcOpenShell#4933]
Now that #4727 is nearing completion it's time to look ahead. I think everyone's plates are pretty full though and its good to let the dust settle a bit.

I would propose we aim for a Google Summer of Code project proposal on authoring alignment/infra geometries in BlenderBIM for the next GSoC season.

This means two simple things:
- We need to find mentors. I'm secretly hoping we can get @civilx64 and @RickBrice to mentor. Mentoring a GSoC student means a couple (could be just 1) hours / week voluntary coaching of a student for a timeframe of (on average) 12 weeks. 
- We need to have a brief project proposal outlined by Jan/Feb 2025. See [0] for a similar proposal.

So this is all rather relaxed at this time.

[0] https://github.com/opencax/GSoC/issues/94

**Comment by civilx64** [comment]
Yes, I can assist.  I've started to think through infra authoring in BBIM conceptually (and probably jumped the gun with some comments in #4727 :smiley: ) so the natural next step is to layout out a project approach and start moving forward.  GSoC is a good idea but if we set it up as a project here on GH we could potentially get other community contributions as well prior to summer 2025.

**Comment by RickBrice** [comment]
Count me in. A few months back I set a reminder to inquire about GSoC. I'd like to get the civil engineering program at my local university involved but don't know where to start

**Comment by aothms** [comment]
> the civil engineering program at my local university involved but don't know where to start

There is no involvement required (or even possible) from universities. Only individuals (no longer required to be students) and mentoring open source organisations. Also the amount of slots is really limited. We can also do a hackathon.

> get other community contributions as well prior to summer 2025.

Definitely! :)

**Comment by RickBrice** [comment]
https://github.com/opencax/GSoC/issues/63 looks like a related example

**Comment by Alx-88** [comment]
Hello everyone,
Any news on implementing alignments?
I'm not a developer (I'm learning python), but I'd like to be part of this topic as I work in the infrastructure sector.

**Comment by Moult** [comment]
@Alx-88 BTW this is about authoring / editing - the viewing of alignments is mostly already implemented.

**Comment by hannesw** [comment]
Hello everyone,

I'm a civil engineer based in Germany, and I'm passionate about developing an open-source alternative for infrastructure design based on IFC. As a starting point, I've expanded on the outline shared by @aothms with some ideas for a minimum viable product (MVP), based on my experiences. I'm excited to collaborate with other professionals to refine this draft further.

# Outline

Alignments are an essential part of designing roads and railways, acting as the framework for planning and building. They include **horizontal alignments**, which show the path’s curves and straight sections when viewed from above (plan view), and **vertical alignments**, which show how the path goes up and down in elevation. Important design factors, like points of intersection (PI), curve radii, slopes, and banking (superelevation), help shape these alignments to make them safe and efficient. Cross-sections, which are slices of the road or railway at regular points, are then swept along the 3D path created by the alignments. This process builds the full design, allowing for features like road widening or slope changes to fit smoothly into the terrain.

Efficient design and modeling require dedicated tools due to the parametric nature of these components. Unlike building models, which have clear spatial references like rooms, walls, and slabs, alignments lack inherent reference points within the model itself. This makes it harder to orient and work directly in the design space. Therefore, the first priority should be to develop foundational tools specifically for creating and managing alignments effectively.

# Expected Outcome

- Create and edit alignments: Both horizontal and vertical, using design parameters for PI.
- Combine alignments: Merge them to form a 3D curve.
- Create and edit cross-sections: Design detailed profiles.
- Extrude cross-sections along the curve: Attach the cross-sections to the curve for extrusion, incorporating a parameter for transition lengths (starting with simple linear transitions).

# Future Possibilities

- Perform cut-and-fill operations with a digital elevation model (DEM).
- Design and connect roads with junctions (a potentially complex task).
- Modify road features: Add or remove lanes and adjust the width of existing ones.
- Generate drawings: Create plans, vertical alignment sections, and cross-sections.

# Additional Information

- I recently came across the [FreeCAD Trails](https://github.com/joelgraff/freecad.trails) plugin, which offers similar functionality. Although it appears to no longer be updated, it’s worth exploring. The author also discusses the plugin in this [video](https://youtu.be/gXVBIvq_WbE?si=WGRnoCsRlC-lJ4GM).
- Check out this [OSArch wiki article](https://wiki.osarch.org/index.php?title=IFC_-_Industry_Foundation_Classes/IFC_alignment) (authored, I believe, by @civilx64 and @RickBrice) for an introduction to alignments.
- Learn from videos about commercial solutions like [Civil 3D](https://www.youtube.com/watch?v=gI80QmABkNU), [OpenRoads](https://www.youtube.com/playlist?list=PLnJUnxLwu_N5JrRr5YUo0-3zyK0vrvT83), or [Allplan Road Design](https://www.youtube.com/playlist?list=PLdOdniujLua0e9JOXYZ-NJfjlhvlAOvFF).
- [Python Example](https://github.com/civilx64/dev-ifcopenshell-4x3-infra/blob/90607ded1a07a6582009626129f8de4475861454/examples/FHWA_Bridge_Geometry_Manual/Alignment.py) by @civilx64

**Comment by aothms** [comment]
@hannesw that's great to hear! welcome!

Currently @RickBrice is the furthest in that he has some working Blender UI code. It seems we're all a bit on and off at the moment, that's a fact of life especially in open source. @RickBrice, do you think we can create a repository around your UI sketch and also list the 10 most immediate small goals to accomplish so that people can just pick things up as they get familiar?

**Comment by RickBrice** [comment]
@hannesw welcome - I’m very excited to have more involvement with alignment because for me it’s a stepping stone to my goal of modeling bridges. 

@civilx64 and I have a planning project that might help. https://github.com/orgs/IfcOpenShell/projects/18

@hannesw there is a brainstorm in https://github.com/IfcOpenShell/IfcOpenShell/issues/5638 similar to your. 

I have some prototype UI in a branch called alignment_modeling_experiments. I’m trying to learn how Blender, its UI, and IFCOS work together. I have not been successful - I need to put in more hours. 

@civilx64 has ported some of the C++ alignment building utilities I wrote over to python. 

Things are temporarily more off than on right now with US holiday and work commitments.

**Comment by Moult** [comment]
Just a heads up, I know nothing about civil modeling, but if anybody here is an expert at civil modeling and knows the desired workflow and user interface needed to be built, and is also willing to write the code for Bonsai, I'm happy to have a call with them - they can describe what is needed, and I can write half the code / the skeleton to show them how Blender UI code works and where code is organised and which of our tool functions can be reused / etc, and we can tag-team the effort :)

Ping @RickBrice who has started on this already :)

**Comment by RickBrice** [comment]
@Moult in my world as a bridge engineer, alignments are defined by someone else so I don't know that workflow. However, once the alignment is defined, it needs to be input to model a bridge. That's what I'm working on as a stop-gap measure to have some minimal alignment editing capabilities. 

Hoping to get a student or two from my local university to propose on the GSoC project https://github.com/opencax/GSoC/issues/98. If that comes to be, a call with you would be super helpful.

In the meantime, I could use some help with a UI to position relative to an alignment (IfcAxis2PlacementLinear.Location, with Location=IfcPointByDistanceExpression.DistanceAlong/BasisCurve/Offset

**Comment by Moult** [comment]
@RickBrice sounds good, did you want to jump on a call? I'm in the Sydney timezone.

**Comment by RickBrice** [comment]
@Moult not a good time to jump on a call at the moment. I'm in Seattle time zone -8 hours

You have my direct email. Let's get something scheduled. Tomorrow after 2pm local time is good for me

**Comment by FraJoMen** [comment]
Good evening, everyone,

I would be happy to contribute to this topic. I have prepared a very small presentation where I highlight what I believe are the key aspects of infrastructure modeling [Mod1Din3D.pptx](https://github.com/user-attachments/files/18870962/Mod1Din3D.pptx).

I hope those more experienced in the field will forgive me for greatly simplifying the concepts, but perhaps it can serve as a starter to start working on this topic.

@hannesw, thank you for your comments — really interesting! The project you highlighted, [FreeCAD Trails](https://github.com/joelgraff/freecad.trails), is really interesting.

Do you think we could involve Joel Graff (@joelgraff) and Hakan Seven (@HakanSeven12)— even just to hear about their experience?

**Comment by HakanSeven12** [comment]
Hello everyone,

I read the discussion on alignment and infrastructure modeling in BlenderBIM with great interest. I’d like to share that I’m developing an open-source workbench for FreeCAD called Road Workbench (successor of Trails and Geomatics Workbenches), which focuses on road design. (and later infrastructure modeling)

Road Workbench provides tools for:

    Geopoints ✔️
    Terrain ✔️
    Alignment ✔️
    Profile ✔️
    Structure ✔️
    Superelevation ❌
    Road 🔄
    Regions ❌
    Sections ❌
    Volume ❌
    GeoLine ❌
    LandXML ❌

It is still in development, but I’m working on improving its capabilities and expanding its features. If anyone is interested in road design in an open-source environment, I’d be happy to discuss it further.

You can find the project here: https://github.com/HakanSeven12/Road

Looking forward to any feedback or discussions!

**Comment by aothms** [comment]
@FraJoMen I think is a nice overview. Would be good to add the IFC-specific vocabulary to this as it helps people get up to speed quicker and the good news is the majority of these is fully implemented; i.e for curves this mean evaluate distance along curve into a 4x4 matrix so that we can build these sweeps. Have a look at https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcGradientCurve.htm https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSegmentedReferenceCurve.htm https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcProfileDef.htm https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSectionedSolidHorizontal.htm (there are more; but as @RickBrice said in the past, the writing in the ifc spec sometimes tends to a have a bit of after happy hour intoxication to it).

@HakanSeven12 Happy to hear you continued your efforts! What are your thoughts on IFC? In the BIM workbench there is some interoperability between Bonsai/BlenderBIM - FreeCAD due to the ability to serialize concepts as IFC and/or OpenCASCADE geometries. Of course everybody is busy and actual interoperability on every front would slow everything down, but the fact that there are similar technological foundations underneath each of them I find really reassuring.

**Comment by HakanSeven12** [comment]
Thanks! IFC could indeed be a fundamental format for my work. I’ve been exploring the Python implementation, particularly this module:
[IfcOpenShell Alignment Module](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.8.0/src/ifcopenshell-python/ifcopenshell/alignment.py).

For Road Workbench, proper IFC integration would require:

Support for clothoid spirals in alignment,
Two different profile definitions: one extracted from a surface and one based on an alignment,
Support for parabolic and asymmetric parabolic vertical curves,
Incorporating superelevation definitions within the alignment structure.
For reference on alignment geometry, you can check out:
[OpenRoadEd](https://github.com/fhwedel-hoe/OpenRoadEd)
Or take a look at my workbench—it’s not perfect, but it gets the job done!

Until now, I personally prefer the LandXML documentation:
[LandXML 1.2 Schema](http://www.landxml.org/schema/landxml-1.2/documentation/landxml-1.2doc.html)

maybe it can help.

**Comment by aothms** [comment]
> Support for clothoid spirals in alignment,

Check. IfcClothoid and all other transition curves in IFC are implemented as curve segments types in a way to evaluate distance along into a 4x4 matrix that allows to render a tessellated curve, perform linear placement and build sweeps.

> Two different profile definitions: one extracted from a surface and one based on an alignment,

We have nothing for surface yet, but I guess some kind of topography - alignment interaction is planned to be investigated in our GSoC project (assuming we find an applicant for it).

>  Support for parabolic and asymmetric parabolic vertical curves,

From the IFC spec we read:

> parabolic arc segment - [IfcAlignmentVerticalSegmentTypeEnum](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcAlignmentVerticalSegmentTypeEnum.htm) .PARABOLICARC. which can describe symmetric parabolas, unsymmetric parabolas can be created by defining two symmetric parabolas segments

So it's not supported per se, but can be worked around. Is that an impediment? The way this works in IFC is that you have infinite curve or spiral and a definition that trims that into a segment.

>  Incorporating superelevation definitions within the alignment structure.

Check. That's https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSegmentedReferenceCurve.htm (there is still a PR open to improve the docs https://github.com/buildingSMART/IFC4.3.x-development/pull/913)

---

Anyway that's where we stand; no pressure to adopt anything. As part of the GSoC we'll also do some investigation of the possibilities for an integrated freecad-bonsai workflow. Especially since we seem to have approached the problem space from somewhat complementary ends - where you have topography and authoring and we maybe have more of the BIM workflow.

**Comment by HakanSeven12** [comment]
I'm interested. One more question—LandXML is widely used in my workflow. I wonder if there's any existing transformation method between LandXML and IFC, or if any work has been done in that direction?

**Comment by aothms** [comment]
yes, something like that exists, but I'm having trouble finding it. 

@peterrdf Is your LandXML to IFC converter not publicly available under https://github.com/bSI-RailwayRoom/IFC-Rail-Unit-Test-Reference-Code ?

**Comment by civilx64** [comment]
I have a python package for LandXML parsing that is complete for loading horizontal, vertical, and cant with auto-generated data structures for schema versions 1.0, 1.1, 1.2, and the draft 2.0. The repo is private at the moment but I am open to suggestions on the best method for making it public

- Part of IfcOpenShell organization as separate repo
- Part of IfcOpenShell repo as separate package like ifctester
- independent of IFCOS and FreeCAD

**Comment by HakanSeven12** [comment]
For me, if I can use it as an integrated part of IfcOpenShell, that would be more useful. So second option looks good to me.

**Comment by civilx64** [comment]
Okay - plan will be to house it at `/src/ifcopenshell-python/landxml/`.  Note this is a "plan", not a "promise" 😉 given other constraints and priorities.

**Comment by peterrdf** [comment]
Dear all, you can find some example code here: https://rdf.bg/download/LandXML2IFC4.3-20240507.zip

**Comment by civilx64** [comment]
> Dear all, you can find some example code here: https://rdf.bg/download/LandXML2IFC4.3-20240507.zip

Many thanks - this will be invaluable for debugging and unit testing!
